src/csexp.ml

```diff
@@ -29,107 +37,135 @@ module Make (Sexp : Sexp) = struct
     let int_of_digit c = Char.code c - Char.code '0'
 
     let rec parse_atom input len =
-      match Input.read_char input with
-      | '0' .. '9' as c ->
+      Input.Monad.bind (Input.read_char input) @@ function
+      | Result.Ok c when '0' <= c && c <= '9' ->
         let len = (len * 10) + int_of_digit c in
         if len > Sys.max_string_length then
-          parse_error "atom too big to represent"
+          Input.Monad.return @@ parse_error "atom too big to represent"
         else
           parse_atom input len
-      | ':' ->
-        let s = Input.read_string input len in
-        Atom s
-      | _ -> invalid_character ()
-
-    let rec parse_many input depth acc =
-      match Input.read_char input with
-      | '(' ->
-        let sexps = parse_many input (depth + 1) [] in
-        parse_many input (depth + 1) (List sexps :: acc)
-      | ')' ->
+      | Result.Ok ':' -> (
+        Input.Monad.bind (Input.read_string input len) @@ function
```

Comment:
Could you bind `Input.Monad.bind` to `>>=` at toplevel and use `>>=` throughout the code? Just to make the code a bit more idiomatic.

Note: you can also add `(preprocess future_syntax)` to the `dune` file to use the `let+` syntax and keep the code compatible with OCaml < 4.08. You'll also need to add a dependency on `ocaml-syntax-shims` for OCaml < 4.08, which is fine by me.